### PR TITLE
Implement template inheritance

### PR DESCRIPTION
**Pull request by jorgepiloto**
Opening this in favor of #16 after experiencing really weird issues with GitHub.

**Main goals of this PR:**
- [x] Allow template inheritance.
- [x] Allow developers to easily add new templates.
- [x] Better modularization of the package.

**Before merging:**  
- [x] Update `pre-commit-config.yml` to work with the new project layout.
- [x] Main CI should check for style, docs, smoke-tests and build of the `ansys-templates` tool.
- [x]  One action per family template whose main goal is to check baked project integrity.
- [x] Add tool code coverage.
- [x] Update documentation by including new guidelines on how to add a new template.
- [x] Add CLI tests.
- [x] Update all README.rst files.
- [x] ~~Clean commit history.~~

### CODE REVIEW

**Comment by jorgepiloto** [comment]
I'm trying to setup the code coverage but it is not ignoring some of the files from the templates even if forcing it to do so.

**Comment by jorgepiloto** [comment]
I just added code coverage! Let me add a `test_cli.py` taking advantage of [cli.testing](https://click.palletsprojects.com/en/8.0.x/testing/) too.

**Comment by jorgepiloto** [comment]
Almost ready for a review. Just have to review all README.rst files. Hopefully, this will be the latest heavy PR of the project.

**Comment by jorgepiloto** [comment]
This is ready for a review. Pinging here @akaszynski and @jgd10 to give feedback on the current implementation. 

It is a heavy PR. Nevertheless, everything has been implemented to allow users to add new templates and reuse existing ones. 

Once this is ready, we could start working on #10.

**Comment by akaszynski** [comment]
Don't worry about clean commit history. Squash and merge, keep commit history in the github PR. Helps with debugging CI failures.

File: doc/source/user_guide/templating.rst [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+   │    │   └── {{cookiecutter.__project_name_slug}}
+   │    │       └── ...
+   │    └── template_N/
+   │        ├── cookiecutter.json
+   │        └── {{cookiecutter.__project_name_slug}}
+   │            └── ...
+   └── family_N/
+
+
+Therefore, you must check if there is already a family for the template you want
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:47]:
```suggestion
Be sure to check if there is already a family for the template you want
```

File: doc/source/user_guide/templating.rst [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+              └── ... # All these files will be included in each family template
+
+Inside this new family, you will need to create at least two different files:
+
+- A ``cookiecutter.json`` file for specifying the minimum required variables for
+  any of the templates of the family to work. You must use cookiecutter private
+  variables in here.
+
+- A ``common/{{cookiecutter.__project_name_slug}}/`` directory. This directory
+  must contain all the typical files for a given family. All files in the
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:74]:
```suggestion
  contains all the typical files for a given family. All files in the
```

File: doc/source/user_guide/templating.rst [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+  ``common/{{cookiecutter.__project_name_slug}}`` directory will be copied to
+  each template when rendering it.
+  
+.. note::
+
+    You can later add a ``hooks/pos_gen_project.py`` file in the
+    ``family/template/`` directory for removing non desired files coming from
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:81]:
```suggestion
    ``family/template/`` directory for removing non-desired files coming from
```

File: doc/source/user_guide/templating.rst [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    the ``common/{{cookiecutter.__project_name_slug}}`` directory.
+
+
+Adding a New Template to a Family
+---------------------------------
+
+To add a new template, start by creating a new template folder. For example, let
+us name it ``src/ansys/templates/family_0/new_template``:
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:89]:
```suggestion
To add a new template, start by creating a new template folder. For example, let's
name it ``src/ansys/templates/family_0/new_template``:
```

File: doc/source/user_guide/templating.rst [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+Therefore, you must check if there is already a family for the template you want
+to add or you need to create one.
+
+
+Adding a New family of Templates
+--------------------------------
+
+Start by creating a new directory in ``src/ansys/templates/`` with the name of
+the new family of templates. For example, let us name it ``src/ansys/templates/new_family``:
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:55]:
```suggestion
the new family of templates. For example, let's name it ``src/ansys/templates/new_family``:
```

File: doc/source/user_guide/templating.rst [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+- A ``{{cookiecutter.__project_name_slug}}/`` directory. This directory must
+  contain any additional files or directories you would like to include in your
+  new template. Notice this folder will be combined with the
+  ``common/{{cookiecutter.__project_name_slug}}/`` one.
+
+
+Adding it to the CLI
+^^^^^^^^^^^^^^^^^^^^
+
+To have access to the template from the CLI (i.a. command line tool), you will
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:119]:
```suggestion
To have access to the template from the CLI (command line interface), you will
```

File: doc/source/user_guide/templating.rst [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+   .. code:: python
+
+       @new.command()
+       def template_name():
+           """Short description of the template."""
+           bake_template(TEMPLATE_PATH_FINDER["pyansys"], os.getcwd())
+
+
+Adding Unitary Tests
+""""""""""""""""""""
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:140]:
```suggestion
Adding Unit Tests
"""""""""""""""""
```

File: doc/source/user_guide/templating.rst [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+- ``tests/tests_templates_family/test_template_family_name_of_template.py``
+
+.. note::
+
+   If you created a new family template, make sure to include tests for the
+   ``family/common/`` directory too.
+
+Expected common files should be included under a
+``tests/tests_templates_family/conftest.py`` as a `pytest fixture`_. As an
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:153]:
```suggestion
``tests/tests_templates_family/conftest.py`` as a `pytest fixture`_. For
```

File: doc/source/user_guide/templating.rst [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+example, consider the following code of a generic ``conftest.py`` file:
+
+.. code::
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:156]:
```suggestion
.. code:: python
```

File: doc/source/user_guide/templating.rst [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+        # Combine all files and export those to be accessible by the tests
+        all_common_files = basedir_files + doc_files + tests_files
+        return all_common_files
+
+
+Add the Family to Tox envs
+""""""""""""""""""""""""""
+
+If you created a new faimly, make sure to add it to the [tox] set of
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:174]:
```suggestion
If you created a new family, make sure to add it to the [tox] set of
```

File: doc/source/user_guide/templating.rst [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+Each family of templates is tested within its own `GitHub actions`_ workflow.
+Therefore, you need to create a:
+
+- ``.github/workflows/templates_family.yml``
+
+.. note::
+
+   To reduce the amount of CI jobs, templates are only tested under Linux based
+   OS. If you require from any particular programming language, try to test the
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:197]:
```suggestion
   To reduce the amount of CI jobs, templates are only tested under Linux.
   If you require from any particular programming language, try to test the
```

File: doc/source/user_guide/templating.rst [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+  variables in here.
+
+- A ``common/{{cookiecutter.__project_name_slug}}/`` directory. This directory
+  must contain all the typical files for a given family. All files in the
+  ``common/{{cookiecutter.__project_name_slug}}`` directory will be copied to
+  each template when rendering it.
+  
+.. note::
+
+    You can later add a ``hooks/pos_gen_project.py`` file in the
```
Comment (akaszynski) [inline doc/source/user_guide/templating.rst:80]:
```suggestion
    You can later add a ``hooks/post_gen_project.py`` file in the
```

File: doc/source/user_guide/usage.rst [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+---------------------------
+
+You can list all available templates by running:
+
+.. code:: bash
+
+   ansys-templates list
+
+Which outputs:
+
```
Comment (akaszynski) [inline doc/source/user_guide/usage.rst:41]:
```suggestion
```

File: doc/source/user_guide/usage.rst [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+.. code:: text
+
+   Available templates in ansys-templates are:
+
+   pybasic: Create a baic Python Package.
```
Comment (akaszynski) [inline doc/source/user_guide/usage.rst:47]:
```suggestion
   pybasic: Create a basic Python Package.
```

File: doc/source/user_guide/usage.rst [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+   pyansys: Create a PyAnsys Python Package project.
+   pyansys_advanced: Create an advanced PyAnsys Python Package project.
+
```
Comment (akaszynski) [inline doc/source/user_guide/usage.rst:50]:
```suggestion
```

File: doc/source/user_guide/usage.rst [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+Creating a New Project
+----------------------
+
+You can create a new project using a given template by running ``ansys-templates
+new`` and passing the name of the template you wish to use.
+
+.. code:: bash
+
+   ansys-templates new template_name
```
Comment (akaszynski) [inline doc/source/user_guide/usage.rst:61]:
```suggestion
   ansys-templates new <template_name>
```

File: doc/source/user_guide/usage.rst [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+   ansys-templates new pybasic
+
+Check available templates names using ``ansys-templates list`` command or just
+run:
+
+.. code:: bash
+
+   ansys-templates new --help
+
+For further information about how to use this command.
```
Comment (akaszynski) [inline doc/source/user_guide/usage.rst:76]:
```suggestion
for further information about how to use this command.
```

File: src/ansys/templates/python/pyansys/README.rst [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+        # Activate it in a Linux environment
+        python -m venv .venv && source .venv/bin/activate
+
+        # Activate it in a Windows CMD environment
+        source .venv\Scripts\activate.bat
+
+        # Activate it in a Windows Powershell environment
+        source .venv\Scripts\Activate.ps1
+
+
```
Comment (akaszynski) [inline src/ansys/templates/python/pyansys/README.rst:53]:
```suggestion
```

File: src/ansys/templates/python/pyansys/README.rst [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+        python -m pip install --editable {{ cookiecutter.__pkg_name }}
+
+5. Install additional requirements (if needed):
+
+     .. code:: bash
+
+        python -m pip install -r requirements/requirements_build.txt
+        python -m pip install -r requirements/requirements_doc.txt
+        python -m pip install -r requirements/requirements_tests.txt
+
```
Comment (akaszynski) [inline src/ansys/templates/python/pyansys/README.rst:73]:
```suggestion
```

File: src/ansys/templates/python/pyansys/README.rst [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+
+6. Finally, verify your development installation by running:
+
+    .. code:: bash
+        
+        python -m pip install -r requirements/requirements_tests.txt
+        pytest tests -vvv
```
Comment (akaszynski) [inline src/ansys/templates/python/pyansys/README.rst:80]: Using higher verbosity levels (-vvv, -vvvv, …) is supported, but has no effect in pytest itself at the moment, however some plugins might make use of higher verbosity.

from:
https://docs.pytest.org/en/latest/how-to/output.html

```suggestion
        pytest tests -v
```

File: src/ansys/templates/python/pyansys/README.rst [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+
+Style and Testing
+-----------------
+
+If required, you can always call the style commands (`black`_, `isort`_,
+`flake8`_...) or unit testing ones (`pytest`_) from the command line. However,
+this does not guarantee that your project is being tested in an isolated
+environment, which is the reason why tools like `tox`_ exist.
```
Comment (akaszynski) [inline src/ansys/templates/python/pyansys/README.rst:89]:
```suggestion
environment, which is another reason to consider using `tox`_.
```

File: src/ansys/templates/python/pyansys/README.rst [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+Documentation
+-------------
+
+For building documentation, you can either run the usual rules provided in the
+`Sphinx`_ Makefile, such us:
+
+.. code:: bash
+
+    python -m pip install -r requirements/requirements_doc.txt
+    make -C doc/ html && your_browser_name doc/html/index.html
```
Comment (akaszynski) [inline src/ansys/templates/python/pyansys/README.rst:101]:
```suggestion
    make -C doc/ html

    # subsequently open the documentation with (under Linux):
    your_browser_name doc/html/index.html
```

**Review by akaszynski — COMMENTED** [review]
Half way done. Feel free to implement the comments in "batch mode" unless I get to it first.

File: doc/source/user_guide/templating.rst (hunk 12 above)
Comment (jorgepiloto) [reply doc/source/user_guide/templating.rst:80]: Good catch!

**Comment by akaszynski** [comment]
Next time, please accept several suggestions in a "batch." See:
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request

File: src/ansys/templates/python/pybasic/README.rst [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+        python -m pip install -r requirements/requirements_doc.txt
+        python -m pip install -r requirements/requirements_tests.txt
+
+
+6. Finally, verify your development installation by running:
+
+    .. code:: bash
+        
+        python -m pip install -r requirements/requirements_tests.txt
+        pytest tests -vvv
```
Comment (akaszynski) [inline src/ansys/templates/python/pybasic/README.rst:77]:
```suggestion
        pytest tests -vv
```

File: src/ansys/templates/python/pybasic/README.rst [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+
+
+Style and Testing
+-----------------
+
+If required, you can always call the style commands (`black`_, `isort`_,
+`flake8`_...) or unit testing ones (`pytest`_) from the command line. However,
+this does not guarantee that your project is being tested in an isolated
+environment, which is the reason why tools like `tox`_ exist.
```
Comment (akaszynski) [inline src/ansys/templates/python/pybasic/README.rst:86]:
```suggestion
environment, which is another reason to use tools like `tox`_.
```

File: src/ansys/templates/python/pybasic/README.rst [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+Documentation
+-------------
+
+For building documentation, you can either run the usual rules provided in the
+`Sphinx`_ Makefile, such us:
+
+.. code:: bash
+
+    python -m pip install -r requirements/requirements_doc.txt
+    make -C doc/ html && your_browser_name doc/html/index.html
```
Comment (akaszynski) [inline src/ansys/templates/python/pybasic/README.rst:98]:
```suggestion
    make -C doc/ html

    # optionally view the generated documentation (on linux) with
    your_browser_name doc/html/index.html
```

File: src/ansys/templates/testing.py [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    Assert if template renders properly.
+
+    Parameters
+    ----------
+    template_path: ~pathlib.Path
+        Path to the project template.
+    output_path: ~pathlib.Path
+        Path to the output baked project.
+    cookiecutter_vars: dict
```
Comment (akaszynski) [inline src/ansys/templates/testing.py:17]:
```suggestion
    template_path : ~pathlib.Path
        Path to the project template.
    output_path : ~pathlib.Path
        Path to the output baked project.
    cookiecutter_vars : dict
```

File: src/ansys/templates/testing.py [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+
+def assert_files_in_baked_project(files_list, project_path):
+    """
+    Assert if given files exists inside desired output path.
+
+    Parameters
+    ----------
+    files_list: list
+        A list of expected files path relative to the output project path.
+    project_path: ~pathlib.Path
```
Comment (akaszynski) [inline src/ansys/templates/testing.py:52]:
```suggestion
    files_list : list
        A list of expected files path relative to the output project path.
    project_path : ~pathlib.Path
```

File: src/ansys/templates/utils.py [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+
+def _copytree(input_path, output_path):
+    """
+    Recursively copy all the contents of a directory into desired one.
+
+    Parameters
+    ----------
+    input_path: ~pathlib.Path
+        Path of the source directory to be copied.
+    output_path: ~pathlib.Path
```
Comment (akaszynski) [inline src/ansys/templates/utils.py:20]:
```suggestion
    input_path : ~pathlib.Path
        Path of the source directory to be copied.
    output_path : ~pathlib.Path
```

File: src/ansys/templates/utils.py [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+
+def _copy_common_template_files(common_path, project_path):
+    """
+    Copy common template files into desired project directory.
+
+    Parameters
+    ----------
+    common_path: ~pathlib.Path
+        Path to the common template directory.
+    project_path: ~pathlib.Path
```
Comment (akaszynski) [inline src/ansys/templates/utils.py:52]:
```suggestion
    common_path : ~pathlib.Path
        Path to the common template directory.
    project_path : ~pathlib.Path
```

File: src/ansys/templates/utils.py [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+
+def _copy_all_template_files(template_path, project_path):
+    """
+    Copy all template files including cookiecutter.json and hooks/ directory.
+
+    Parameters
+    ----------
+    template_path: ~pathlib.Path
+        Path to the template directory.
+    project_path: ~pathlib.Path
```
Comment (akaszynski) [inline src/ansys/templates/utils.py:70]:
```suggestion
    template_path : ~pathlib.Path
        Path to the template directory.
    project_path : ~pathlib.Path
```

File: src/ansys/templates/utils.py [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+
+def _include_license(license_path, project_path):
+    """
+    Include a desired license into the baked project.
+
+    Parameters
+    ----------
+    license_path: ~pathlib.Path
+        Path to the license template.
+    project_path: ~pathlib.Path
```
Comment (akaszynski) [inline src/ansys/templates/utils.py:88]:
```suggestion
    license_path : ~pathlib.Path
        Path to the license template.
    project_path : ~pathlib.Path
```

File: src/ansys/templates/testing.py (hunk 26 above)
Comment (jorgepiloto) [reply src/ansys/templates/testing.py:17]: You are right on this white space between variable name and the semicolon. I though no space was required but [from numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html#parameters) it is clear that this space is required.

**Comment by jorgepiloto** [comment]
> Next time, please accept several suggestions in a "batch." See: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request

Will use it from now on, for sure ☝🏽

File: src/ansys/templates/python/pyansys/README.rst [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+
+2. Create a fresh-clean Python environment and activate it. Refer to the
+   official `venv` documentation if you require further information:
+
+    .. code:: bash
+
+        # Create a virtual environment
+        python -m venv .venv
+
+        # Activate it in a Linux environment
```
Comment (jorgepiloto) [inline src/ansys/templates/python/pyansys/README.rst:44]: This is solves https://github.com/pyansys/pyansys-templates/issues/17 too.

File: src/ansys/templates/python/pyansys/README.rst [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+        python -m venv .venv && source .venv/bin/activate
+
+        # Activate it in a Windows CMD environment
+        source .venv\Scripts\activate.bat
+
+        # Activate it in a Windows Powershell environment
+        source .venv\Scripts\Activate.ps1
```
Comment (jorgepiloto) [inline src/ansys/templates/python/pyansys/README.rst:51]: These lines solve issue #17.

File: src/ansys/templates/python/pyansys/README.rst [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+need to follow these steps:
+
+1. Start by cloning this repository:
+
+    .. code:: bash
+
+        git clone {{ cookiecutter.repository_url }}
+
+2. Create a fresh-clean Python environment and activate it. Refer to the
+   official `venv` documentation if you require further information:
```
Comment (akaszynski) [inline src/ansys/templates/python/pyansys/README.rst:37]:
```suggestion
   official `venv <https://docs.python.org/3/library/venv.html>`_ documentation if you require further information:
```

Comment (akaszynski) [reply src/ansys/templates/python/pyansys/README.rst:37]: Requires the updated link at the bottom of this `README.rst`.

File: src/ansys/templates/python/pyansys/README.rst [hunk 35]
```diff
@@ -0,0 +1,131 @@
...
+.. LINKS AND REFERENCES
+.. _black: https://github.com/psf/black
+.. _flake8: https://flake8.pycqa.org/en/latest/
+.. _isort: https://github.com/PyCQA/isort
+.. _PyAnsys Developer's guide: https://dev.docs.pyansys.com/
+.. _pre-commit: https://pre-commit.com/
+.. _pytest: https://docs.pytest.org/en/stable/
+.. _Sphinx: https://www.sphinx-doc.org/en/master/
+.. _tox: https://tox.wiki/
+
```
Comment (akaszynski) [inline src/ansys/templates/python/pyansys/README.rst:131]:
```suggestion
.. venv: https://docs.python.org/3/library/venv.html

```

File: src/ansys/templates/python/pybasic/README.rst [hunk 36]
```diff
@@ -0,0 +1,132 @@
...
+guide`_. You will need to follow these steps:
+
+1. Start by cloning this repository:
+
+    .. code:: bash
+
+        git clone {{ cookiecutter.repository_url }}
+
+2. Create a fresh-clean Python environment and activate it. Refer to the
+   official `venv` documentation if you require further information:
```
Comment (akaszynski) [inline src/ansys/templates/python/pybasic/README.rst:34]:
```suggestion
   official `venv`_ documentation if you require further information:
```

**Review by akaszynski — APPROVED** [review]
A few minor changes remain, but otherwise this is good to go.

Overall, this is excellent work. I especially like the testing sections, as this looks much easier to maintain via `pytest` rather than the makefile/script used in [ansys/ansys-api-template](https://github.com/ansys/ansys-api-template)

**Comment by jorgepiloto** [comment]
This is finally ready! Thanks @akaszynski  for the review. Merging... 🚀

**Pull request merged**